### ISSUE TITLE
feat: helper functions

### ISSUE THREAD

**Issue by usamoi** [tensorchord/pgvecto.rs#255]
pgvecto.rs are lack of helper functions, and we need to provide more helper functions

* function `vector_norm`: get length of a vector
* function `vector_dimensions`: get dimensions of a vector
* function `vector_normalize`: compute a normalized vector
* casts between `vector` and `vecf16`
* casts between `vecf16` and `half precision`

view https://docs.pgvecto.rs/reference/schema.html to see helper functions provided by pgvecto.rs so far

**Comment by VoVAllen** [comment]
Test logical replication between different extension version

**Comment by JoePassanante** [comment]
Can you elaborate on 
>  casts between vecf16 and half precision; PostgreSQL do not support half precision, we may make some decisions here

Is this an issue with postgres itself?  or that it hasn't been implememted in pgvecto?

**Comment by usamoi** [comment]
> Is this an issue with postgres itself?

Yes. PostgreSQL supports floating-type `real` (IEEE 754 binary 32) and floating type `double precision` (IEEE 754 binary 64), but there is not such a type named `half precision` (IEEE 754 binary 16). We probably cannot provide a cast between a `vecf16` and an array. Another choice is providing a cast between `vef16` and `real[]`, which is "make some decisions here" refered.

**Comment by JoePassanante** [comment]
> > Is this an issue with postgres itself?
> 
> Yes. PostgreSQL supports floating-type `real` (IEEE 754 binary 32) and floating type `double precision` (IEEE 754 binary 64), but there is not such a type named `half precision` (IEEE 754 binary 16). We probably cannot provide a cast between a `vecf16` and an array. Another choice is providing a cast between `vef16` and `real[]`, which is "make some decisions here" refered.

Messed over in the discord for some help getting the environment setup.  But isn't this as simple as:

```rust
#[pgrx::pg_extern(immutable, parallel_safe, strict)]
fn _vectors_cast_vecf32_to_vecf16(
    vector: Vecf32Input<'_>,
    _typmod: i32,
    _explicit: bool,
) -> Vecf16Output {
    let data = vector
        .data()
        .iter()
        .map(|x| x.to_f32())
        .map(|x| f16::from_f32(x))
        .collect();

    Vecf16::new_in_postgres(data)
}

CREATE CAST (vector AS vecf16)
    WITH FUNCTION _vectors_cast_vecf32_to_vecf16(vector, integer, boolean) AS IMPLICIT;
```

**Comment by usamoi** [comment]
> > > Is this an issue with postgres itself?
> > 
> > 
> > Yes. PostgreSQL supports floating-type `real` (IEEE 754 binary 32) and floating type `double precision` (IEEE 754 binary 64), but there is not such a type named `half precision` (IEEE 754 binary 16). We probably cannot provide a cast between a `vecf16` and an array. Another choice is providing a cast between `vef16` and `real[]`, which is "make some decisions here" refered.
> 
> Messaged over in the discord for some help getting the environment setup. But isn't this as simple as:
> 
> ```rust
> #[pgrx::pg_extern(immutable, parallel_safe, strict)]
> fn _vectors_cast_vecf32_to_vecf16(
>     vector: Vecf32Input<'_>,
>     _typmod: i32,
>     _explicit: bool,
> ) -> Vecf16Output {
>     let data = vector
>         .data()
>         .iter()
>         .map(|x| x.to_f32())
>         .map(|x| f16::from_f32(x))
>         .collect();
> 
>     Vecf16::new_in_postgres(data)
> }
> 
> CREATE CAST (vector AS vecf16)
>     WITH FUNCTION _vectors_cast_vecf32_to_vecf16(vector, integer, boolean) AS IMPLICIT;
> ```

Yes.

> * casts between vector and vecf16
> * casts between vecf16 and half precision[]

It's two issues. You can talk only about the first one since you only need the first one.

**Comment by VoVAllen** [comment]
I think float(2) is the same as fp16?

**Comment by usamoi** [comment]
> I think float(2) is the same as fp16?

https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-FLOAT

I don't think so.

**Comment by my-vegetable-has-exploded** [comment]
Is there more specific definition of `absolute-value norm`?

**Comment by usamoi** [comment]
> Is there more specific definition of `absolute-value norm`?

https://en.wikipedia.org/wiki/Norm_(mathematics)

**Comment by my-vegetable-has-exploded** [comment]
> > Is there more specific definition of `absolute-value norm`?
>
> https://en.wikipedia.org/wiki/Norm_(mathematics)

I did not grasp the definition. How should we calculate it for a one-dimensional vector?

**Comment by usamoi** [comment]
> > > Is there more specific definition of `absolute-value norm`?
> > 
> > 
> > https://en.wikipedia.org/wiki/Norm_(mathematics)
> 
> I did not grasp the definition. How should we calculate it for a one-dimensional vector?

https://baike.baidu.com/item/%E5%90%91%E9%87%8F%E7%9A%84%E6%A8%A1/2073854
